### ISSUE TITLE
Deploy validator clients

### ISSUE THREAD

**Issue by narimiran** [status-im/infra-nimbus#122]
This is a continuation of #111 ("(...) if the setup proves to be stable enough, we might deploy some validators to the consensus node.")

This can either be a new role or an expansion of a current role.
This should be done for all testnets.
We should start with a relatively small number of validators initially, e.g. on the nodes that currently run 60 validators.

This can be done in one of these two ways:
- New boolean flag `use_validator_client`: indicating that all validators on that host should be attached to the validator client.
- Alternatively: allow the hosts to have validators attached both to the beacon node and to the validator client. New property `number_of_validators`: the number of validators attached to a validator client.

More information about running the validator client is available [here](https://nimbus.guide/validator-client.html).

**Comment by jakubgs** [comment]
I assume since you want this enabled on nodes with 60 validators that would be `metal-05.he-eu-hel1.nimbus.prater`:
https://github.com/status-im/infra-nimbus/blob/ddc8de5fa46adfdf4ef04b2956a8007ae5657bbe/ansible/group_vars/nimbus.prater.yml#L101-L105
On all branches or just some?

**Comment by jakubgs** [comment]
Before this issue can be resolved we first need a proper dedicated EL client node setup, as right now all beacon nodes are using the same Geth node running on `goerli-01.aws-eu-central-1a.nimbus.geth`.

* https://github.com/status-im/infra-nimbus/issues/125

**Comment by jakubgs** [comment]
I started some work on this. Or at least I did some thinking:

* I will set this up as part of the existing `infra-role-beacon-node-*` roles to avoid too much boilerplate.
* I'm going to make use of the `--secrets-dir` and `--validators-dir` flags and point them at `/var/empty` or similar.
* I will leave the secrets and validators in the same place as they already are, except they will be served by the client.

I will probably get a version working on Monday.

**Comment by jakubgs** [comment]
Changes necessary to manage location of `secrets` and `validators` folders separately:

- https://github.com/status-im/infra-role-dist-validators/commit/6d7d9e37 - provide secrets and validators paths separately
- https://github.com/status-im/infra-role-beacon-node-linux/commit/3e8e553f - configure secrets and validators paths explicitly
- https://github.com/status-im/infra-role-beacon-node-macos/commit/b51df7bf - configure secrets and validators paths explicitly
- https://github.com/status-im/infra-role-beacon-node-windows/commit/ac7696da - configure secrets and validators paths explicitly
- https://github.com/status-im/infra-nimbus/commit/928d5a12 - requirements: specify validator paths explicitly

Now we can point those at `/var/empty` when using a validator client.

**Comment by arnetheduck** [comment]
>     I'm going to make use of the --secrets-dir and --validators-dir flags and point them at /var/empty or similar.
>    I will leave the secrets and validators in the same place as they already are, except they will be served by the client.

eh this feels risky - one restart where the flag is misspelled or some other shit reason, and the validators are gone

**Comment by jakubgs** [comment]
>eh this feels risky - one restart where the flag is misspelled or some other shit reason, and the validators are gone

That's the point. They are supposed to be gone.

**Comment by jakubgs** [comment]
Based on discussion with @arnetheduck and reading of the main issue:

* https://github.com/status-im/nimbus-eth2/issues/3088

It seems it makes sense to create a separate Ansible role makes the most sense, since we might want to do 1-N setups.

**Comment by jakubgs** [comment]
I've created a repo for the separate ansible role: https://github.com/status-im/infra-repos/commit/764a8225

https://github.com/status-im/infra-role-validator-client

I will be basing it mostly on `infra-role-beacon-node-linux`.

**Comment by jakubgs** [comment]
Some initial work:

- https://github.com/status-im/infra-role-validator-client/commit/14939d96 - add basic Ansible metadata
- https://github.com/status-im/infra-role-validator-client/commit/7a7eb52f - add initial setup of folders and buinary build

**Comment by jakubgs** [comment]
More changes to get this going:

- https://github.com/status-im/infra-role-validator-client/commit/d22e685f - add task for configuring systemd service
- https://github.com/status-im/infra-role-validator-client/commit/1a049ecb - add task for deploying validators
- https://github.com/status-im/infra-role-validator-client/commit/ebf86bf9 - add task for opening ports on firewall
- https://github.com/status-im/infra-role-validator-client/commit/cc48f725 - add task for configuring consul services
- https://github.com/status-im/infra-role-validator-client/commit/eb681bc1 - enable keymanager endpoint by default

**Comment by jakubgs** [comment]
This is interesting:
```json
{"lvl":"NOT","ts":"2022-09-23 08:28:15.955+00:00","msg":"Starting REST HTTP server","url":"http://127.0.0.1:5053/"}
{"lvl":"INF","ts":"2022-09-23 08:28:15.956+00:00","msg":"Beacon node has been identified","agent":"Nimbus/v22.9.1-72e6b2-stateofus","service":"fallback_service","endpoint":"127.0.0.1:9301"}
{"lvl":"INF","ts":"2022-09-23 08:28:15.956+00:00","msg":"Beacon node has compatible configuration","service":"fallback_service","endpoint":"127.0.0.1:9301 [Nimbus/v22.9.1-72e6b2-stateofus]"}
{"lvl":"INF","ts":"2022-09-23 08:28:15.956+00:00","msg":"Beacon node is in sync","sync_distance":0,"head_slot":833241,"is_opimistic":"false","service":"fallback_service","endpoint":"127.0.0.1:9301 [Nimbus/v22.9.1-72e6b2-stateofus]"}
{"lvl":"NOT","ts":"2022-09-23 08:28:15.957+00:00","msg":"Fork schedule updated","fork_schedule":[{"previous_version":"0x80000069","current_version":"0x80000069","epoch":0},{"previous_version":"0x80000069","current_version":"0x80000070","epoch":500},{"previous_version":"0x80000070","current_version":"0x80000071","epoch":750}],"service":"fork_service"}
{"lvl":"ERR","ts":"2022-09-23 08:28:27.959+00:00","msg":"Unable to get head state's validator information","service":"duties_service"}
{"lvl":"NOT","ts":"2022-09-23 08:28:27.961+00:00","msg":"REST service started","address":"127.0.0.1:5053"}
{"lvl":"INF","ts":"2022-09-23 08:28:27.961+00:00","msg":"Scheduling first slot action","startTime":"16w3d17h28m27s961ms398us840ns","nextSlot":833243,"timeToNextSlot":"8s38ms601us160ns"}
{"lvl":"INF","ts":"2022-09-23 08:28:27.962+00:00","msg":"Beacon node has been identified","agent":"Nimbus/v22.9.1-72e6b2-stateofus","service":"fallback_service","endpoint":"127.0.0.1:9301 [Nimbus/v22.9.1-72e6b2-stateofus]"}
{"lvl":"INF","ts":"2022-09-23 08:28:27.963+00:00","msg":"Beacon node has compatible configuration","service":"fallback_service","endpoint":"127.0.0.1:9301 [Nimbus/v22.9.1-72e6b2-stateofus]"}
{"lvl":"INF","ts":"2022-09-23 08:28:27.964+00:00","msg":"Beacon node is in sync","sync_distance":1,"head_slot":833241,"is_opimistic":"false","service":"fallback_service","endpoint":"127.0.0.1:9301 [Nimbus/v22.9.1-72e6b2-stateofus]"}
{"lvl":"WRN","ts":"2022-09-23 08:28:29.985+00:00","msg":"Connection with beacon node(s) has been lost","online_nodes":0,"unusable_nodes":1,"total_nodes":1,"service":"fallback_service"}
{"lvl":"WRN","ts":"2022-09-23 08:28:31.970+00:00","msg":"No suitable beacon nodes available","online_nodes":0,"offline_nodes":1,"uninitalized_nodes":0,"incompatible_nodes":0,"nonsynced_nodes":0,"total_nodes":1,"service":"fallback_service"}
```
It seems to be failing in a loop of:

1. `Beacon node has been identified`
2. `Beacon node has compatible configuration`
3. `Beacon node is in sync`
4. `Connection with beacon node(s) has been lost`
5. `No suitable beacon nodes available`

Over and over again. @narimiran any idea?

**Comment by jakubgs** [comment]
I can also see that our usual healthcheck for Consul isn't there in Keymanager API:
```
admin@metal-01.he-eu-hel1.nimbus.ropsten:~ % c http://localhost:5053/eth/v1/node/version
curl: (22) The requested URL returned error: 404 Not Found
```
So for now I'm going to use a TCP healthcheck, but it would be nice to have a route without auth for this.

**Comment by jakubgs** [comment]
Some more changes:

- https://github.com/status-im/infra-role-validator-client/commit/4200da56 - user: fix creating sudoers files
- https://github.com/status-im/infra-role-validator-client/commit/dedbdcaa - build: fix repo path variable name
- https://github.com/status-im/infra-role-validator-client/commit/73e3e4c0 - service: drop unnecessary --era-dir flag
- https://github.com/status-im/infra-role-validator-client/commit/763972bb - config: create keymanager token file and use it
- https://github.com/status-im/infra-role-validator-client/commit/928073b7 - consul: use tcp healthcheck for keymanager API

And a fix for massive headers sent to the beacon node:

- https://github.com/status-im/infra-role-beacon-node-linux/commit/f04d6a46 - service: specify reset max body and max headers size

Which was causing this:
```json
{
  "lvl": "ERR",
  "ts": "2022-09-23 11:12:24.002+00:00",
  "msg": "Unable to get head state's validator information",
  "service": "duties_service"
}
```

**Comment by jakubgs** [comment]
And here's the setup on `nimbus.ropsten`: https://github.com/status-im/infra-nimbus/commit/7d05abad

**Comment by jakubgs** [comment]
Lowered number of validators for VC nodes as requested by @cheatfate:

- https://github.com/status-im/infra-nimbus/commit/b1760d9a - nimbus.ropsten: use less validators on VC nodes

https://github.com/status-im/infra-nimbus/blob/b1760d9ab647ebcb1b29f4a453c673e027a0c91e/ansible/group_vars/nimbus.ropsten.yml#L83-L87

**Comment by jakubgs** [comment]
I forgot to note o friday that the beacon nodes with validator clients connected are using obscene amounts of memory:

![image](https://user-images.githubusercontent.com/2212681/192300977-ef6de2ac-ac33-4f3b-85e8-0ec97289a5d2.png)

The ones without validator clients are fine though.

**Comment by jakubgs** [comment]
I also deployed validator client for the first node on the Sepolia host: https://github.com/status-im/infra-nimbus/commit/70f3da73
https://github.com/status-im/infra-nimbus/blob/70f3da73d2843f84b5dc9efc35e0f450b616c8a8/ansible/group_vars/nimbus.sepolia.yml#L76-L80

And also implemented disabling service and its checks:

- https://github.com/status-im/infra-role-consul-service/commit/4849a3c9 - add disabled attribute to check definition
- https://github.com/status-im/infra-role-validator-client/commit/4acb2262 - service: allow disabling the service by default

**Comment by jakubgs** [comment]
And interestingly enough on Sepolia no such memory issues appear:

![image](https://user-images.githubusercontent.com/2212681/192726364-123e2f46-5dcf-4a35-8b65-9fa77ce133f4.png)

![image](https://user-images.githubusercontent.com/2212681/192726695-d1c61737-622e-4406-b6fe-9ff5af9da1b3.png)

Although that might be a function of number of validators attached. Or maybe network-specific.

**Comment by jakubgs** [comment]
I also deployed validator client for the `stable` node on `linux-04` host on mainnet: https://github.com/status-im/infra-nimbus/commit/269a76a2
https://github.com/status-im/infra-nimbus/blob/269a76a23ee5d9a76ab8b0e787c9fb2bddf42f82/ansible/group_vars/nimbus.prater.yml#L147-L151
And it appears to run fine and without memory issues so far, but the doppelganger detection is taking an awful long time:
```
admin@linux-04.he-eu-hel1.nimbus.prater:~ % grep 'Attestation has not been served' /data/validator-client-prater-stable-01/logs/service.log | head -n1
{"lvl":"INF","ts":"2022-10-11 11:36:28.002+00:00","msg":"Attestation has not been served (doppelganger check still active)","slot":4081682,"validator":"94cab382","validator_index":307765,"service":"attestation_service"}

admin@linux-04.he-eu-hel1.nimbus.prater:~ % grep 'Attestation has not been served' /data/validator-client-prater-stable-01/logs/service.log | tail -n1
{"lvl":"INF","ts":"2022-10-11 12:20:04.002+00:00","msg":"Attestation has not been served (doppelganger check still active)","slot":4081900,"validator":"94cad9d2","validator_index":310801,"service":"attestation_service"}
```
Already 45 minues and still going. Appears to be a bug.

**Comment by jakubgs** [comment]
For now I've disabled doppelganger for the validator temporarily: https://github.com/status-im/infra-nimbus/commit/5de20671https://github.com/status-im/infra-nimbus/blob/5de206719e12e7b3e29364cc8474f46625a7cb1e/ansible/group_vars/nimbus.prater.yml#L85
Once the fix is merged I'll undo that.

**Comment by jakubgs** [comment]
Looks like we are in business:
```
admin@linux-04.he-eu-hel1.nimbus.prater:~ % grep 'Attestation published' /data/validator-client-prater-stable-01/logs/service.log | tail -n4
{"lvl":"NOT","ts":"2022-10-11 12:28:16.005+00:00","msg":"Attestation published","attestation":{"aggregation_bits":"0x000000000000000000000000000000000000000000080020","data":{"slot":4081941,"index":62,"beacon_block_root":"84c6728e","source":"127559:5e90d514","target":"127560:2208b6af"},"signature":"abef638f"},"validator":"94cad9d2","validator_index":310801,"delay":"5ms15us884ns","service":"attestation_service"}
{"lvl":"NOT","ts":"2022-10-11 12:28:28.057+00:00","msg":"Attestation published","attestation":{"aggregation_bits":"0x000000000000000800000000000000000000000000000020","data":{"slot":4081942,"index":43,"beacon_block_root":"97f79d1a","source":"127559:5e90d514","target":"127560:2208b6af"},"signature":"b52d9f94"},"validator":"94cc88e5","validator_index":306447,"delay":"57ms822us424ns","service":"attestation_service"}
{"lvl":"NOT","ts":"2022-10-11 12:29:52.019+00:00","msg":"Attestation published","attestation":{"aggregation_bits":"0x000000000000000000000000000000000000000040000020","data":{"slot":4081949,"index":29,"beacon_block_root":"665b69cd","source":"127559:5e90d514","target":"127560:2208b6af"},"signature":"a3764bad"},"validator":"94cab382","validator_index":307765,"delay":"19ms516us82ns","service":"attestation_service"}
{"lvl":"NOT","ts":"2022-10-11 12:31:04.006+00:00","msg":"Attestation published","attestation":{"aggregation_bits":"0x004000000000000000000000000000000000000000000020","data":{"slot":4081955,"index":60,"beacon_block_root":"39e0176f","source":"127560:2208b6af","target":"127561:79b5ab0a"},"signature":"ad20a7b6"},"validator":"94c79ff7","validator_index":304582,"delay":"6ms122us158ns","service":"attestation_service"}
```
But one things that makes me wonder is why the validator client message is `Attestation published` while the beacon node message is `Attestation sent`. Seems like an unnecessary divergence that can just cause confusion.

**Comment by jakubgs** [comment]
This is currently working on all 3 testnets: `sepolia`, `ropsten`, and `prater`
https://github.com/status-im/infra-nimbus/blob/c8e3232dff2f92d6b40e92b89dfc92941a41e45e/ansible/group_vars/nimbus.sepolia.yml#L76-L77
https://github.com/status-im/infra-nimbus/blob/c8e3232dff2f92d6b40e92b89dfc92941a41e45e/ansible/group_vars/nimbus.ropsten.yml#L82-L84
https://github.com/status-im/infra-nimbus/blob/c8e3232dff2f92d6b40e92b89dfc92941a41e45e/ansible/group_vars/nimbus.prater.yml#L148-L149
I consider this done. Reopen if there's something missing.

**Comment by jakubgs** [comment]
Based on suggestion from @arnetheduck I've also deployed a VC for `unstable` on `linux-03`: https://github.com/status-im/infra-nimbus/commit/cf8bab14
